### PR TITLE
Support blocking untrusted operations

### PR DESCRIPTION
**Pull request by atjn**
Support blocking untrusted and arbitrary operations with [`vips_block_untrusted_set`](https://www.libvips.org/API/current/libvips-vips.html#vips-block-untrusted-set) and [`vips_operation_block_set`](https://www.libvips.org/API/current/VipsOperation.html#vips-operation-block-set):

```js
// Block all untrusted operations
vips.blockUntrusted(true);

// Block all foreign loaders
vips.operationBlock("VipsForeignLoad", true);

// Reenable the jpeg loader
vips.operationBlock("VipsForeignLoadJpeg", true);
```

Note that it was already possible to use `vips_block_untrusted_set` before, by setting the `module.ENV.VIPS_BLOCK_UNTRUSTED` parameter on initialization.

Related: https://github.com/kleisauke/wasm-vips/pull/21#issuecomment-1221549162

### CODE REVIEW

**Comment by atjn** [comment]
I considered integrating this a bit deeper, allowing for things like:

```js
vips.Operation.block(vips.Image.pngsave, true);

vips.Image.jpegsave.block(true);

vips.Image.jpegsave.block(); // returns true
vips.Image.jpegsave.untrusted(); // returns false

```

But AFAICT it would require a massive amount of work, and I wasn't even sure if it would be desirable, given that it is pretty different from the core `libvips` API.

File: build/preamble_vips.d.ts
Comment (kleisauke) [inline build/preamble_vips.d.ts:84]: I think it would be clearer to just provide a full JavaScript example, for example:

```suggestion
    /**
     * Block a specific operation in the libvips class hierarchy from running.
     * For example:
     * ```js
     * vips.operationBlock('VipsForeignLoad', true);
     * vips.operationBlock('VipsForeignLoadJpeg', false);
     * ```
     * Will block all load operations, except JPEG. Use:
     * ```bash
     * $ vips -l
     * ```
     * at the command-line to see the class hierarchy.
     * @param name The name of the operation in the libvips class hierarchy.
     * @param state Set to `true` to block the operation, set to `false` to re-enable it.
     */
```

File: src/vips-emscripten.cpp
Comment (kleisauke) [inline src/vips-emscripten.cpp:438]: Perhaps we could use `optional_override` to avoid the need of a separate `init.h` file?
```suggestion
    function("blockUntrusted", optional_override([](bool state) {
                 vips_block_untrusted_set(state ? 1 : 0);
             }));
    function("operationBlock",
             optional_override([](const std::string &name, bool state) {
                 vips_operation_block_set(name.c_str(), state ? 1 : 0);
             }));
```

File: test/unit/helpers.js
Comment (kleisauke) [inline test/unit/helpers.js:241]: I think this logic should probally be moved to `test_block.js` instead. For example, you could do this:
```js
  let cachePreviousMax;

  before(function () {
    // Temporary disable the operation cache of libvips to avoid caching blocked operations
    cachePreviousMax = vips.Cache.max();
    vips.Cache.max(0);
  });

  after(function () {
    // Re-enable the operation cache of libvips
    vips.Cache.max(cachePreviousMax);
  });

  afterEach(function () {
    cleanup();
  });
```
(fwiw, the `afterEach`-block is not really necessary, but helps in debugging ref leaks and ensures that the images are properly cleaned up after every test)

File: test/unit/test_block.js
Comment (kleisauke) [inline test/unit/test_block.js:28]: This test is duplicated, so I think these lines can be removed.
```suggestion
```

File: test/unit/test_block.js
Comment (kleisauke) [inline test/unit/test_block.js:6]: Let's use the camelCase operation as name here.
```suggestion
  it('blockUntrusted', function () {
```

File: test/unit/test_block.js
Comment (kleisauke) [inline test/unit/test_block.js:32]: Ditto.
```suggestion
  it('operationBlock', function () {
```

**Review by kleisauke — COMMENTED** [review]
Thanks for working on this! I left a couple of comments inline.

File: test/unit/test_block.js
Comment (atjn) [reply test/unit/test_block.js:28]: It depends on how thorough you want to test it. The idea is to test that it still blocks `vipsLoad` even after `vipsLoad` has been specifically greenlit and has been called. I felt like there could be muddy waters here, and I was _sort of, but not really_ right, because it actually doesn't pass this test when the cache is enabled.

File: src/vips-emscripten.cpp
Comment (atjn) [reply src/vips-emscripten.cpp]: Sure, moved in https://github.com/kleisauke/wasm-vips/pull/24/commits/97f3648831a2deb1f7e67f48bdd935661ff93a67

File: test/unit/test_block.js
Comment (kleisauke) [reply test/unit/test_block.js:28]: > It depends on how thorough you want to test it. The idea is to test that it still blocks `vipsLoad` even after `vipsLoad` has been specifically greenlit and has been called.

Ah, I see. Makes sense.

> it actually doesn't pass this test when the cache is enabled.

Oh curious, let me investigate that. I think the `operation is blocked`-error should always be thrown immediately and not be kept in the cache. Perhaps explicitly calling `.avg()` on the blocked operation would still cause that error?

/cc @jcupitt

Comment (atjn) [reply test/unit/test_block.js:28]: I already mentioned this for John, and he argued that since the `vipsLoad` method is never called, there is no problem serving the loaded image from cache.

I see it both ways. I think both approaches have upsides and downsides.

Comment (jcupitt) [reply test/unit/test_block.js:28]: Hi Kleis, I suppose when an operation is blocked we could search the cache and knock any operations of that type out. 

It wouldn't add security, since cached operations are not executed, but I suppose it might make behaviour less confusing. It's a cosmetic change heh.

I guess we could also check for blocked status on cache lookup, that might be simpler, and would avoid any threading issues.

File: test/unit/helpers.js
Comment (atjn) [reply test/unit/helpers.js]: Makes sense to me. I thought you could only apply those functions globally to all running tests.
Moved in https://github.com/kleisauke/wasm-vips/pull/24/commits/6dcf7925ea43ef89d3837cb3bee55c61dcb3e9e8

Comment (kleisauke) [reply test/unit/helpers.js]: I initially meant to move it to the `describe` block, but this also seems to work, TIL.

File: test/unit/test_block.js
Comment (kleisauke) [reply test/unit/test_block.js:42]: I misunderstood, sorry. I thought that blocked operations were cached, causing the error to be displayed only once, but this is not the case.

IIUC, the issue is that cached operations are not knocked out of the cache when `vips_block_untrusted_set( TRUE );` is called? I suppose this is a issue. For example, revisiting the test like this:
```js
// By default, libvips doesn't block untrusted operations
expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.not.throw();

// However, if the environment variable `VIPS_BLOCK_UNTRUSTED` is set,
// or `vips_block_untrusted_set( TRUE );` is called, any operations
// that we've tagged as untrusted will be prevented from running
vips.blockUntrusted(true);

// For example, vipsload is an untrusted operation, and should throw
// when untrusted operations are blocked
// FIXME: Though, the first `vipsload` call will store the operation
// in cache, so subsequent calls will originate from their
expect(() => vips.Image.vipsload(Helpers.vipsFile)).to.throw(/operation is blocked/);

// Ensure no operations are blocked when the rest of the tests are run
vips.blockUntrusted(false);
```

Would cause a test failure.

> I guess we could also check for blocked status on cache lookup -- that might be simpler, and would avoid any threading issues.

Good idea. Another (simpler) idea is to just call `vips_cache_drop_all();` when `vips_block_untrusted_set( TRUE );` is called, to ensure that cached operations are invalidated.

File: test/unit/helpers.js
Comment (kleisauke) [reply test/unit/helpers.js]: (I moved it into the `describe` block with commit ef0088e01814414bc6d6c2e10846952b7f7823af to be consistent with the other tests)

File: test/unit/test_block.js
Comment (kleisauke) [reply test/unit/test_block.js]: ... though, that issue shouldn't block this PR, I just pushed commit ef0088e01814414bc6d6c2e10846952b7f7823af and will merge once the CI is green. :smile:

Comment (jcupitt) [reply test/unit/test_block.js]: I think it'd be simplest to add it here:

https://github.com/libvips/libvips/blob/master/libvips/iofuncs/cache.c#L801

Just swap that line for:

```C
                if( hit->invalid ||
                        (VIPS_OPERATION_GET_CLASS( hit->operation )->flags &
                         VIPS_OPERATION_BLOCKED) ) {
                        /* Has been tagged for removal, or has been blocked.
                         */
                        vips_cache_remove( hit->operation );
                        hit = NULL;
                }
```

It should be fast, and no race conditions are possible.

Comment (kleisauke) [reply test/unit/test_block.js]: Nice! Feel free to incorporate that, I think it would be useful for other bindings as well.

Comment (kleisauke) [reply test/unit/test_block.js]: ... I confirm that this change works,  just pushed as commit 0fdd86bb820416562d9618a633e70f269174c98d, thanks!

Comment (jcupitt) [reply test/unit/test_block.js]: OK, pushed to master with https://github.com/libvips/libvips/commit/702ed8298f45d7ba342ebf5bae612d159e9cec6f

Comment (kleisauke) [reply test/unit/test_block.js]: Great! I think we're done, I'll mark this conversation as resolved.

**Review by kleisauke — APPROVED** [review]
Thanks all! This will be in v0.0.4.

**Pull request merged**
